### ISSUE TITLE
GeneralizedNewtypeDeriving broken on GHC 7.8.2

### ISSUE THREAD

**Issue by tazjin** [acid-state/safecopy#12]
Hej,

I noticed this today when trying to build my blog with GHC 7.8.2

Minimal complete example:

``` haskell
{-# LANGUAGE GeneralizedNewtypeDeriving #-}

import Data.SafeCopy

newtype Foo = Foo Integer
            deriving (SafeCopy)
```

Output:

```
~/s/acid-test> ghc Broken.hs 
[1 of 1] Compiling Main             ( Broken.hs, Broken.o )

Broken.hs:6:23:
    Could not coerce from ‘Kind Integer’ to ‘Kind Foo’
      because the first type argument of ‘Kind’ has role Nominal,
      but the arguments ‘Integer’ and ‘Foo’ differ
      arising from the coercion of the method ‘kind’ from type
                   ‘Kind Integer’ to type ‘Kind Foo’
    Possible fix:
      use a standalone 'deriving instance' declaration,
        so you can specify the instance context yourself
    When deriving the instance for (SafeCopy Foo)
```

Using the TH call deriveSafeCopy works fine, but seems to generate instances incompatible with earlier instances derived automatically. Fails with an error: `Could not parse saved checkpoint due to the following error: Failed reading: safecopy: Integer: Cannot find getter associated with this version number: Version {unVersion = 5131390}`

Any ideas?

//V

**Comment by MichaelXavier** [comment]
@dschalk @tazjin I'm a bit confused. WIll switching to standalone deriving not work? I've got a lot of code that used GND and I'm trying to figure out if i should use standalone instance or TH without having to go through a major migration headache.

**Comment by sdroege** [comment]
@MichaelXavier during my testing standalone deriving did not work for all instances, and for those it did it changed the representation somehow to make a migration necessary nonetheless. In the end I just used deriveSafeCopy for every type and migrated them all to that then.

**Comment by MichaelXavier** [comment]
@sdroege since "migrated" is an overloaded term in this context I want to clarify: using standalone instances created in some cases byte-incompatible safecopy whereas code that was previously using GND that you changed to deriveSafeCopy always produced byte-compatible safecopy?

**Comment by sdroege** [comment]
@MichaelXavier No, sorry. Both deriveSafeCopy and standalone instances created byte-incompatible instances, but standalone instances did not work for all types for me (got compiler errors for some). So I settled with using deriveSafeCopy for everything and migrated the old GND instances to the new byte-incompatible deriveSafeCopy instances.

**Comment by MichaelXavier** [comment]
Jeeze. We could really use some comment from the package authors. This can cause a whole lot of damage. I'm in the process of adding safecopy tests to all of my projects that use it and I guess hand-writing dozens of safecopy instances.

**Comment by ozataman** [comment]
Would really love to hear comments from the package authors on this. This is holding up our migration of a large code base from 7.6 to 7.8. We have dozens of newtypes with GND and it would be a major headache to manually migrate them.  Even if we put in the work, what's even scarier is the potential for mistakes and data corruption in production.

**Comment by Lemmih** [comment]
I'm not sure what kind of comment you want. Either someone writes a bit of TH code to mimic GND or you'll have to write the instances by hand. It sucks that you can no longer use GND but it's not the end of the world.

**Comment by MichaelXavier** [comment]
Can the library provide a migration path for this? Thankfully 7.8 raises a compile time error when this is encountered, so the user has to stop and think. It would be much better if the docs for this library had a section on this and provided a function that mimicked the GND. The alternative is the user guesses the right thing to do (standalone instance derive according to GHC's suggestion) and breaks safecopy on all their newtypes silently.

**Comment by Lemmih** [comment]
Yes, I would like that very much. Since several people (including you) are affected by this, surely we won't have to wait too long for a pull request. :)

**Comment by Lemmih** [comment]
I quite interested in when standalone deriving isn't the same as ordinary deriving. It would be helpful if @ozataman added a test case.

**Comment by MichaelXavier** [comment]
@ozataman and I work together. Part of our upgrade involves getting a test suite around safecopy instances to test this. We haven't had the time yet to get back to that project the last week or so. @sdroege since you have experienced this incompatibility firsthand, do you think you could produce an example of standalone producing different output? Maybe its confined to certain types that were GNDed.

**Comment by sdroege** [comment]
See this commit for the case where it broke for me (just a little toy project in case someone is wondering):
https://github.com/sdroege/duckduckbot/commit/d5a2f79e7f26ce0f14077570cf43d115d93fa5cd
deriveSafeCopy worked on both but broke the serialization format, standalone deriving caused a compiler error (with ghc 7.8.3)

I won't have time to look into this in the next days, so if it's urgent for anybody please take those information and try to reproduce it. Otherwise I'll try to have some time for this in the following weeks.

**Comment by Lemmih** [comment]
That is expected. The TH code is very different from GND. The interesting case is when standalone deriving is different from ordinary deriving.

**Comment by sdroege** [comment]
Sure, the problem for me was more that standalone deriving did not compile at all :smile:

**Comment by MichaelXavier** [comment]
@Lemmih I've finally gotten back to this and I can confirm in my case that standalone deriving produces the exact same error: 

```haskell
    Could not coerce from ‘Kind Text’ to ‘Kind TextWrapper’
      because the first type argument of ‘Kind’ has role Nominal,
      but the arguments ‘Text’ and ‘TextWrapper’ differ
      arising from a use of ‘GHC.Prim.coerce’
    In the expression: GHC.Prim.coerce (kind :: Kind Text) :: Kind TextWrapper
    In an equation for ‘kind’:
        kind = GHC.Prim.coerce (kind :: Kind Text) :: Kind TextWrapper
    When typechecking the code for  ‘kind’
      in a standalone derived instance for ‘SafeCopy TextWrapper’:
      To see the code I am typechecking, use -ddump-deriv
    In the instance declaration for ‘SafeCopy TextWrapper’
```

I've been playing around with some ways to generically implement these safecopy instances to delegate to the wrapped type's implementation but because Contained does not export anything and is not a Functor, I can't see a way I could wrap the existing implementation. Instead I have to look up the implementation and then re-implement, inserting the newtype Constructor as needed.

**Comment by andreasabel** [comment]
GHC 7 has been dropped, so I suppose this issue is stale.
